### PR TITLE
feat: improve handling of errors

### PR DESCRIPTION
**Pull request by DNR500**
Jira: [LF-8182](https://lifi.atlassian.net/browse/LF-8182)

[LF-8182]: https://lifi.atlassian.net/browse/LF-8182?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/helpers.ts [hunk 1]
```diff
@@ -62,7 +62,8 @@ export const convertQuoteToRoute = (step: LiFiStep): Route => {
 
 export const fetchTxErrorDetails = async (txHash: string, chainId: number) => {
   const response = await request<TenderlyResponse>(
-    `https://api.tenderly.co/api/v1/public-contract/${chainId}/tx/${txHash}`
+    `https://api.tenderly.co/api/v1/public-contract/${chainId}/tx/${txHash}`,
+    { disableLiFiErrorCodes: true }
```
Comment (DNR500) [inline src/helpers.ts:66]: We use the request function in a couple of places to make requests from sources outside of the LiFi domain - This flag opts out of using the LiFi error codes in situations like this.

File: src/services/api.ts [hunk 2]
```diff
@@ -48,20 +47,15 @@ export const getToken = async (
...
-        chain,
-        token,
-      } as Record<string, string>)}`,
-      {
-        signal: options?.signal,
-      }
-    )
-    return response
-  } catch (e) {
-    throw await parseBackendError(e)
```
Comment (DNR500) [inline src/services/api.ts:63]: parseBackendError as basically do two main things - using response.json() to access the response body message and then also creating an error with sets of error codes and error types dependant on status code of the response.

Both these tasks have been moved into the HTTPError class - with the async method to process the response.json() being triggered in the request function. The try/catch shouldn't be needed here anymore.

File: src/core/EVM/EVMStepExecutor.ts [hunk 3]
```diff
@@ -480,7 +484,7 @@ export class EVMStepExecutor extends BaseStepExecutor {
       })
       this.statusManager.updateExecution(step, 'FAILED')
       console.warn(e)
```
Comment (chybisov) [inline src/core/EVM/EVMStepExecutor.ts:486]: Let's remove this warn

File: src/core/EVM/EVMStepExecutor.ts [hunk 4]
```diff
@@ -87,9 +86,13 @@ export class EVMStepExecutor extends BaseStepExecutor {
...
-      throw new TransactionError(
-        LiFiErrorCode.WalletChangedDuringExecution,
-        errorMessage
+      throw await parseEVMStepErrors(
+        new TransactionError(
+          LiFiErrorCode.WalletChangedDuringExecution,
+          errorMessage
+        ),
+        step,
+        process
```
Comment (chybisov) [inline src/core/EVM/EVMStepExecutor.ts:95]: Do you think it makes sense to parse it here, as this code is usually wrapped inside another try-catch, and we will end up parsing it twice? Also, I think we should only parse errors before throwing them to the top as a last step.

File: src/core/EVM/parseEVMStepErrors.ts
Comment (chybisov) [inline src/core/EVM/parseEVMStepErrors.ts:1]: We discussed naming them as `parseEVMError.ts` and `parseSVMError.ts`/`parseSolanaError.ts`

File: src/helper.unit.spec.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+describe('helpers', () => {
+  describe('checkPackageUpdates', () => {
+    beforeEach(() => {
+      vi.spyOn(global, 'fetch').mockResolvedValue({
+        json: () => Promise.resolve({ version: latestVersion }),
+      } as Response)
+
+      vi.spyOn(console, 'warn').mockImplementation(() => {})
+    })
```
Comment (chybisov) [inline src/helper.unit.spec.ts:22]: Should we move this to msw as well? 😄

File: src/utils/errors/index.ts
Comment (chybisov) [inline src/utils/errors/index.ts:1]: How about we move errors outside of utils to have it's own folder?

**Review by chybisov — CHANGES_REQUESTED** [review]

File: src/utils/errors/index.ts
Comment (DNR500) [reply src/utils/errors/index.ts:1]: I like that idea

File: src/core/EVM/parseEVMStepErrors.ts
Comment (DNR500) [reply src/core/EVM/parseEVMStepErrors.ts:1]: Makes sense

File: src/core/EVM/EVMStepExecutor.ts (hunk 3 above)
Comment (DNR500) [reply src/core/EVM/EVMStepExecutor.ts:486]: Yes its the only console.warn I think. Makes sense to remove it

File: src/helper.unit.spec.ts (hunk 5 above)
Comment (DNR500) [reply src/helper.unit.spec.ts:22]: Can do

Comment (DNR500) [reply src/helper.unit.spec.ts:22]: Thinking about it the latest version endpoint isn't under our control and isn't so likely to change - I think mocking with fetch is fine

File: src/core/EVM/EVMStepExecutor.ts (hunk 4 above)
Comment (DNR500) [reply src/core/EVM/EVMStepExecutor.ts:95]: The checkWallet function isn't always used in a part of the executeStep method that isn't cover by parseEVMStep Errors in the try catch there. I think if we look to simplify the executeStep method future we could look to reduce the use of parseEVMStepErrors then.

**Pull request merged**